File: typecast-php/examples/quick-cloning.php
<?php

/**
 * Quick Voice Cloning — PHP SDK Example
 *
 * Demonstrates how to clone a voice from an audio sample, use it for TTS,
 * and delete it when done.
 *
 * Usage:
 *   TYPECAST_API_KEY=your-key php examples/quick-cloning.php [path/to/sample.wav]
 *
 * If no audio file is specified, a small synthetic WAV is generated in-memory
 * so the example runs without any external files.
 */

require_once __DIR__ . '/../vendor/autoload.php';

use Neosapience\Typecast\Exceptions\TypecastException;
use Neosapience\Typecast\Models\CustomVoice;
use Neosapience\Typecast\Models\TTSRequest;
use Neosapience\Typecast\TypecastClient;

// ---------------------------------------------------------------------- //
// Configuration                                                           //
// ---------------------------------------------------------------------- //

$apiKey = getenv('TYPECAST_API_KEY');
if (!$apiKey) {
    echo "Error: set the TYPECAST_API_KEY environment variable.\n";
    exit(1);
}

$audioPath = $argv[1] ?? null;

// ---------------------------------------------------------------------- //
// Prepare audio bytes                                                     //
// ---------------------------------------------------------------------- //

if ($audioPath !== null) {
    if (!file_exists($audioPath)) {
        echo "Error: file not found — {$audioPath}\n";
        exit(1);
    }
    $audioBytes = file_get_contents($audioPath);
    $filename   = basename($audioPath);
    echo "Using audio file: {$audioPath} (" . strlen($audioBytes) . " bytes)\n";
} else {
    // Generate a minimal valid WAV in memory (44-byte header + 1 s of silence).
    $filename   = 'sample.wav';
    $audioBytes = makeSilentWav(sampleRate: 16000, durationSeconds: 1);
    echo "No audio file provided — using 1-second silent WAV (" . strlen($audioBytes) . " bytes)\n";
}

// Validate size before calling the API.
if (strlen($audioBytes) > CustomVoice::CLONING_MAX_FILE_SIZE) {
    echo "Error: audio exceeds 25 MB limit.\n";
    exit(1);
}

// ---------------------------------------------------------------------- //
// Clone voice                                                             //
// ---------------------------------------------------------------------- //

$client = new TypecastClient(apiKey: $apiKey);

echo "\n--- Cloning voice ---\n";
try {
    $voice = $client->cloneVoice(
        audio: $audioBytes,
        filename: $filename,
        name: 'PHP Example Voice',
        model: 'ssfm-v21',
    );
} catch (\InvalidArgumentException $e) {
    echo "Validation error: {$e->getMessage()}\n";
    exit(1);
} catch (TypecastException $e) {
    echo "API error: {$e->getMessage()}\n";
    exit(1);
}

echo "Cloned voice ID : {$voice->voiceId}\n";
echo "Name            : {$voice->name}\n";
echo "Model           : {$voice->model}\n";

// ---------------------------------------------------------------------- //
// Use the cloned voice for TTS                                            //
// ---------------------------------------------------------------------- //

echo "\n--- Synthesising speech with the cloned voice ---\n";
try {
    $tts = $client->textToSpeech(new TTSRequest(
        voiceId: $voice->voiceId,
        text: 'Hello! This is my quick-cloned voice speaking.',
        model: $voice->model,
    ));

    $outPath = __DIR__ . '/cloned_output.wav';
    file_put_contents($outPath, $tts->audioData);
    echo "Saved TTS output to {$outPath} ({$tts->duration}s)\n";
} catch (TypecastException $e) {
    echo "TTS error: {$e->getMessage()}\n";
    // Attempt cleanup even after TTS failure.
}

// ---------------------------------------------------------------------- //
// Delete the cloned voice                                                 //
// ---------------------------------------------------------------------- //

echo "\n--- Deleting cloned voice ---\n";
try {
    $client->deleteVoice($voice->voiceId);
    echo "Deleted voice {$voice->voiceId} — done.\n";
} catch (TypecastException $e) {
    echo "Delete error: {$e->getMessage()}\n";
    exit(1);
}

echo "\nExample completed successfully.\n";

// ---------------------------------------------------------------------- //
// Helper: build a minimal 1-channel 16-bit PCM WAV in memory             //
// ---------------------------------------------------------------------- //

function makeSilentWav(int $sampleRate, int $durationSeconds): string
{
    $numChannels   = 1;
    $bitsPerSample = 16;
    $numSamples    = $sampleRate * $durationSeconds;
    $dataSize      = $numSamples * $numChannels * ($bitsPerSample / 8);
    $byteRate      = $sampleRate * $numChannels * ($bitsPerSample / 8);
    $blockAlign    = $numChannels * ($bitsPerSample / 8);

    $header  = 'RIFF';
    $header .= pack('V', 36 + $dataSize);   // chunk size
    $header .= 'WAVE';
    $header .= 'fmt ';
    $header .= pack('V', 16);               // subchunk1 size
    $header .= pack('v', 1);                // PCM = 1
    $header .= pack('v', $numChannels);
    $header .= pack('V', $sampleRate);
    $header .= pack('V', $byteRate);
    $header .= pack('v', $blockAlign);
    $header .= pack('v', $bitsPerSample);
    $header .= 'data';
    $header .= pack('V', $dataSize);
    $header .= str_repeat("\x00", $dataSize);

    return $header;
}

File: typecast-zig/examples/quick_cloning.zig
/// Instant cloning example for the Typecast Zig SDK.
///
/// Usage:
///   TYPECAST_API_KEY=<key> zig build example-quick-cloning
///
/// The example reads a WAV file from disk, clones a voice, synthesises a
/// short phrase with the new voice, and finally deletes the custom voice.
const std = @import("std");
const typecast = @import("typecast");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // ── Configuration ────────────────────────────────────────────────────
    const api_key = std.posix.getenv("TYPECAST_API_KEY") orelse {
        std.debug.print("Error: TYPECAST_API_KEY environment variable not set.\n", .{});
        std.process.exit(1);
    };

    const base_url = std.posix.getenv("TYPECAST_API_HOST") orelse "https://api.typecast.ai";

    // Audio file to clone from.  You can replace this with any WAV/MP3 ≤ 25 MB.
    const audio_path = std.posix.getenv("CLONE_AUDIO_PATH") orelse "sample.wav";

    // ── Load audio ───────────────────────────────────────────────────────
    const audio_file = std.fs.cwd().openFile(audio_path, .{}) catch |err| {
        std.debug.print("Error: cannot open audio file '{s}': {}\n", .{ audio_path, err });
        std.process.exit(1);
    };
    defer audio_file.close();

    const audio_bytes = try audio_file.readToEndAlloc(allocator, typecast.CLONING_MAX_FILE_SIZE);
    defer allocator.free(audio_bytes);

    std.debug.print("Loaded audio: {s} ({d} bytes)\n", .{ audio_path, audio_bytes.len });

    // ── Create client ────────────────────────────────────────────────────
    var client = typecast.Client.init(allocator, .{
        .api_key = api_key,
        .base_url = base_url,
    });
    defer client.deinit();

    // ── Clone voice ──────────────────────────────────────────────────────
    std.debug.print("Cloning voice...\n", .{});

    const custom_voice = try client.cloneVoice(
        allocator,
        audio_bytes,
        audio_path,     // filename — used for MIME type detection
        "My Zig Voice", // display name (1–30 chars)
        "ssfm-v30",     // model
    );
    defer {
        allocator.free(custom_voice.voice_id);
        allocator.free(custom_voice.name);
        allocator.free(custom_voice.model);
    }

    std.debug.print("Cloned voice: id={s}  name={s}  model={s}\n", .{
        custom_voice.voice_id,
        custom_voice.name,
        custom_voice.model,
    });

    // ── Synthesise with the new voice ────────────────────────────────────
    std.debug.print("Synthesising with the custom voice...\n", .{});

    const tts_response = try client.textToSpeech(.{
        .voice_id = custom_voice.voice_id,
        .text = "Hello! This audio was created with my cloned Zig voice.",
        .model = .ssfm_v30,
    });
    defer allocator.free(tts_response.audio_data);

    const out_path = "cloned_output.wav";
    const out_file = try std.fs.cwd().createFile(out_path, .{});
    defer out_file.close();
    try out_file.writeAll(tts_response.audio_data);

    std.debug.print("Saved {s} ({d} bytes, {d:.2}s)\n", .{
        out_path,
        tts_response.audio_data.len,
        tts_response.duration,
    });

    // ── Delete the custom voice ──────────────────────────────────────────
    std.debug.print("Deleting custom voice {s}...\n", .{custom_voice.voice_id});
    try client.deleteVoice(custom_voice.voice_id);
    std.debug.print("Done.\n", .{});
}
